examples/f16/components/fgbase.py

```diff
@@ -0,0 +1,218 @@
+        parameters: longitude and latitude (radians)
+        returns: list of quaternion components (scalar last)
+        """
+        zd2 = 0.5 * lon
+        yd2 = -0.25 * math.pi - 0.5 * lat
+        Szd2 = math.sin(zd2)
+        Syd2 = math.sin(yd2)
+        Czd2 = math.cos(zd2)
+        Cyd2 = math.cos(yd2)
+        w = Czd2 * Cyd2
+        x = -Szd2 * Syd2
+        y = Czd2 * Syd2
+        z = Szd2 * Cyd2
+        return [x, y, z, w]
+    
+    @classmethod
+    def convert_to_ecef(cls, pn_m: float, pe_m: float, pu_m: float,
+                        phi_rad: float, theta_rad: float, psi_rad: float,
+                        lat0_rad: float, lon0_rad: float, h0_m: float) -> typing.Tuple[float]:
+        """
+        This method takes in a dictionary of "raw" 2D Dubins log data, as read from the LogReader class,
+        and returns a populated Episode object.
+        """
+        # position conversions
+
+        ## ENU to ECEF
+        # convert position to geocentric (Earth-centered) reference frame
+        ecef_x, ecef_y, ecef_z = pm.enu2ecef(pe_m, pn_m, pu_m, lat0_rad, lon0_rad, h0_m, ell=None, deg=False)
+
+        # orientation conversions
+
+        ## ECEF
+        # 1st rotation (frame alignment)
+        global_rotation = Rotation.from_quat(Dubins2DConverter.quaternion_from_lon_lat(
+            lon0_rad, lat0_rad))
+        # 2nd rotation (from data)
+        local_rotation = Rotation.from_euler('xyz', [phi_rad, theta_rad, psi_rad], degrees=False)
+
+        # multiply
+        rotation = global_rotation * local_rotation
+
+        quaternion = rotation.as_quat()
+        angle = 2 * math.acos(quaternion[3])  # cos(a / 2) = w
+        direction = quaternion / (math.sin(angle / 2))  # [Vx,Vy,Vz] * sin(a / 2) = [x,y,z]
+        ecef_x_orientation = direction[0] * angle
+        ecef_y_orientation = direction[1] * angle
+        ecef_z_orientation = direction[2] * angle
+
+        return (
+            ecef_x, ecef_y, ecef_z,
+            ecef_x_orientation,
+            ecef_y_orientation,
+            ecef_z_orientation
+        )
+
+class FlightGearBase(ABC):
+    # Start position of the aircraft
+    DEFAULT_FG_LAT = 35.802117
+    DEFAULT_FG_LON = -117.806717
+    DEFAULT_FG_GROUND_LEVEL = 715 # m
+    # Default max values for actutors
+    DEFAULT_FG_AILERON_MAX_DEG = 21.5
+    DEFAULT_FG_ELEVATOR_MAX_DEG = 25
+    DEFAULT_FG_RUDDER_MAX_DEG = 30.0
+
+    FG_FT_IN_M = 3.2808
+    # Networking variables
+    DEFAULT_FG_IP = "127.0.0.1"
+
+    DEFAULT_DELTA_T = 0.5
+
+    # Class variables
+    reset_flag = False
+    plant = None
+    controller = None
+
+    lag = DEFAULT_DELTA_T
+    speed = 1.0
+    initial_time = None
+    sim_flag = False
+    stopped = False
+    main_loop = None
+
+    lat0 = np.deg2rad(DEFAULT_FG_LAT)
+    lon0 = np.deg2rad(DEFAULT_FG_LON)
+    h0 = DEFAULT_FG_GROUND_LEVEL
+
+    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP
```

Comment:
I would recommend having a `sock_args = (socket.AF_INET, socket.SOCK_DGRAM)` which is then turned into a socket at initialization `self.sock = socket.socket(*self.sock_args)`. This is because if one imports this script, they don't necessarily assume that an OS resource (endpoint + descriptor) was taken.

Reply:
Fixed